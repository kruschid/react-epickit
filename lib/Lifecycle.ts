import * as React from "react";

interface ILifecycleProps {
  onCreate?: () => any;
  onDestroy?: () => any;
}

export class Lifecycle extends React.Component<ILifecycleProps> {
  constructor(props: any) {
    super(props);
  }

  public componentDidMount() {
    if (this.props.onCreate) {
      this.props.onCreate();
    }
  }

  public componentWillUnmount() {
    if (this.props.onDestroy) {
      this.props.onDestroy();
    }
  }

  public render() {
    return this.props.children || null;
  }
}
