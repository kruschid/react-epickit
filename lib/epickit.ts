import { QueueingSubject } from "queueing-subject";
import { BehaviorSubject, merge, Observable, Subject } from "rxjs";
import { filter, ignoreElements, map, mergeAll, share, tap } from "rxjs/operators";

export type Reducer<S, P = any> = (state: S, payload: P) => S;
export type Epic<S = any, A extends IAction = IAction<S>, R extends IAction = IAction<S>> = (
  action$: Observable<A>,
  state$: Observable<S>,
) => Observable<R>;

export interface IAction <S = any, P = any> {
  type?: symbol;
  payload: P;
  reducer?: Reducer<S, P>;
}

export type DispatchFn<S> = (action: IAction<S>) => void;

export interface IEpicKit<S> {
  state$: BehaviorSubject<S>;
  action$: Subject<IAction<S>>;
  epic$: Observable<[S, IAction<S>]>;
  dispatch: DispatchFn<S>;
}

export const createAction = <S = any>(
  typeOrReducer: symbol | Reducer<S>,
  reducer?: Reducer<S>,
): IAction<S> => ({
  type: typeof typeOrReducer === "symbol" ? typeOrReducer : undefined,
  reducer: typeof typeOrReducer === "function" ? typeOrReducer : reducer,
  payload: undefined,
});

export const createActionWithPayload = <S, P>(
  typeOrReducer: symbol | Reducer<S, P>,
  reducer?: Reducer<S, P>,
) => (
  payload: P,
): IAction<S, P> => ({
  type: typeof typeOrReducer === "symbol" ? typeOrReducer : undefined,
  reducer: typeof typeOrReducer === "function" ? typeOrReducer : reducer,
  payload,
});

export const reduceState = <S, T extends IAction<S>>(
  state$: BehaviorSubject<S>,
  action$: Observable<T>,
): Observable<[S, T]> =>
  action$.pipe(
    tap(({reducer, payload}) => {
      if (reducer) {
        state$.next(reducer(state$.getValue(), payload));
      }
    }),
    map<T, [S, T]>((action) => [state$.getValue(), action]),
  );

export const connectEpics = <S>(
  state$: BehaviorSubject<S>,
  action$: Subject<IAction<S>>,
  epics: Array<Epic<S>>,
): Observable<never> =>
  merge(
    epics.map((epic) => epic(action$, state$)),
  )
  .pipe(
    mergeAll(),
    tap((action) => action$.next(action)),
    ignoreElements(),
  );

export const filterAction = <P, S = any>(type: symbol | symbol[]) => (observable$: Observable<IAction<S, P>>) =>
  observable$.pipe(
    filter((action) =>
      Array.isArray(action.type) ? action.type.indexOf(type) >= 0 : action.type === type,
    ),
  );

export const createEpicKit = <S>(initialState: S, epics: Array<Epic<S>> = []): IEpicKit<S> => {
  const state$ = new BehaviorSubject<S>(initialState);
  const action$ = new QueueingSubject<IAction<S>>();
  const epic$ = merge(
    connectEpics(state$, action$, epics),
    reduceState(state$, action$),
  )
  .pipe(share());

  const dispatch: DispatchFn<S> = (action) =>
    action$.next(action);

  return {
    state$,
    action$,
    epic$,
    dispatch,
  };
};
