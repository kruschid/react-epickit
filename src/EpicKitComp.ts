import * as React from "react";
import { BehaviorSubject, merge, Observable, Subject, Subscription } from "rxjs";
import { filter, mergeAll, share, tap } from "rxjs/operators";

// tslint:disable-next-line
type FirstArgument<T> = T extends (arg1: infer U, ...args: any[]) => any ? U : any;

export type Reducer<S, P = never> = (state: S, payload?: P) => S;

export type Epic<S = any, A extends IAction = IAction<S>, R extends IAction = IAction<S>>
  = (action$: Observable<A>, state$: Observable<S>) => Observable<R>;

export interface IAction <S = any, P = never> {
  type?: symbol;
  payload?: P;
  reducer?: Reducer<S, P>;
}
// type ConnectedDispatch<P = never> = (arg: P) => void;

type DispatchFn<S, P = never> = (action: IAction<S, P>) => void;

type IConnectedActions<L> = {
  [x in keyof L]: (payload: FirstArgument<L[x]>) => void;
};

interface IActionList {
  [x: string]: ReturnType<ICreateAction>;
}

interface IEpicKitCompProps<S, L> {
  initialState?: S;
  epics?: Array<Epic<S>>;
  actions?: L;
  children: (p: {state: S, dispatch: DispatchFn<S>, actions: IConnectedActions<L>}) => React.ReactNode;
}

interface IEpicKit<S> {
  state$: Observable<S>;
  action$: Observable<IAction<S>>;
  epic$: Observable<IAction<S>>;
  dispatch: DispatchFn<S>;
}

interface ICreateAction {
  <S, P = never>(type: symbol, reducer?: Reducer<S>): (payload?: P) => IAction<S, P>;
  <S, P = never>(reducer: Reducer<S>): (payload?: P) => IAction<S, P>;
}

export const createAction: ICreateAction = <S = any, P = never>(
  typeOrReducer: symbol | Reducer<S>,
  reducer?: Reducer<S>,
) => (
  payload?: P,
): IAction<S, P> => ({
  type: typeof typeOrReducer === "symbol"
    ? typeOrReducer as symbol : undefined,
  reducer: typeof typeOrReducer === "function"
    ? typeOrReducer as Reducer<S> : reducer,
  payload,
});

const reduceState = <S, T extends IAction<S>>(
  state$: BehaviorSubject<S>,
  action$: Observable<T>,
): Observable<T> =>
  action$.pipe(
    filter(({reducer}) => !!reducer),
    tap(({reducer, payload}) => state$.next(reducer(state$.getValue(), payload))),
  );

const connectEpics = <S>(
  state$: BehaviorSubject<S>,
  action$: Subject<IAction<S>>,
  epics: Array<Epic<S>>,
): Observable<IAction<S>> =>
  merge(
    epics.map((epic) => epic(action$, state$)),
  )
  .pipe(
    mergeAll(),
    tap((action) => action$.next(action)),
  );

export const EpicKit = <S>(initialState: S, epics: Array<Epic<S>> = []): IEpicKit<S> => {
  const state$ = new BehaviorSubject<S>(initialState);
  const action$ = new Subject<IAction<S>>();
  const epic$ = merge(
    connectEpics(state$, action$, epics),
    reduceState(state$, action$),
  )
  .pipe(share());

  const dispatch: DispatchFn<S> = (action) =>
    action$.next(action);

  return {
    state$,
    action$,
    epic$,
    dispatch,
  };
};

export class EpicKitComp<S, L extends IActionList> extends React.Component<IEpicKitCompProps<S, L>> {
  public state: S;

  private epicKit: IEpicKit<S>;
  private actions: IConnectedActions<L>;
  private queue: Array<IAction<S>>;
  private subscription: Subscription;

  constructor(p: IEpicKitCompProps<S, L>) {
    super(p);

    this.epicKit = EpicKit(p.initialState, p.epics);
    this.state = p.initialState;
    this.actions = Object.keys(p.actions).map((key: keyof L) =>
      (payload: any) => this.dispatch(p.actions[key]()),
    );
  }

  public dispatch(action: IAction<S>) {
    if (this.subscription) {
      this.epicKit.dispatch(action);
    } else {
      this.queue.push(action);
    }
  }

  public componentDidMount() {
    this.subscription = merge(
      this.epicKit.epic$,
      this.epicKit.state$.pipe(tap(this.setState.bind(this))),
    )
    .subscribe();
  }

  public componentWillUnmount() {
    this.subscription.unsubscribe();
  }

  public render() {
    return this.props.children({
      state: this.state,
      dispatch: this.dispatch.bind(this),
      actions: this.actions,
    });
  }
}
