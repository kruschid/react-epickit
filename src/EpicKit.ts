import * as React from "react";
import { Subscription } from "rxjs";

import { createEpicKit, DispatchFn, Epic, IEpicKit } from "../lib/epickit";

export interface IStateDispatchProps<S> {
  state: S;
  dispatch: DispatchFn<S>;
}

export interface IEpicKitProps<S> {
  epicKit?: IEpicKit<S>;
  initialState?: S;
  epics?: Array<Epic<S>>;
  children: (p: IStateDispatchProps<S>) => React.ReactNode;
}
export class EpicKit<S> extends React.Component<IEpicKitProps<S>> {
  public state: S;

  private epicKit: IEpicKit<S>;
  private subscription: Subscription | undefined;

  constructor(p: IEpicKitProps<S>) {
    super(p);

    if (p.epicKit) {
      this.epicKit = p.epicKit!;
      this.state = p.epicKit.state$.getValue();
    } else if (p.initialState) {
      this.epicKit = createEpicKit<S>(p.initialState, p.epics);
      this.state = p.initialState!;
    } else {
      throw Error("EpicKit component requires epicKit or initialState property to be set");
    }
  }

  public componentDidMount() {
    this.subscription = this.epicKit.epic$.subscribe(([state]) =>
      this.setState(state),
    );
  }

  public componentWillUnmount() {
    this.subscription!.unsubscribe();
  }

  public render() {
    return this.props.children({
      state: this.state,
      dispatch: this.epicKit.dispatch,
    });
  }
}
