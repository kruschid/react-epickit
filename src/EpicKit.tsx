import * as React from "react";
import { DispatchFn, IEpicKit } from "epickit";

interface IStateDispatchProps<S> {
  state: S;
  dispatch: DispatchFn<S>;
}

export function EpicKitState<S>(p: {
  epicKit: IEpicKit<S>;
  children: (p: IStateDispatchProps<S>) => React.ReactElement;
}) {
  const [state, setState] = React.useState(p.epicKit.state$.getValue());

  React.useEffect(() => {
    const subscription = p.epicKit.epic$.subscribe(
      ({state}) => setState(state),
    );
    return () => subscription.unsubscribe();
  });

  return p.children({
    state: state,
    dispatch: p.epicKit.dispatch,
  })
}

export const EpicKit = <S, >(epicKit: IEpicKit<S>) => {
  const Context = React.createContext<IStateDispatchProps<S>>({
    state: epicKit.state$.getValue(),
    dispatch: epicKit.dispatch,
  });

  return {
    Consumer: Context.Consumer,
    Provider: (p: {children: React.ReactElement}) =>
      <EpicKitState epicKit={epicKit}>
        {(epicKitProps) =>
          <Context.Provider value={epicKitProps}>
            {p.children}
          </Context.Provider>
        }
      </EpicKitState>
  };
};
