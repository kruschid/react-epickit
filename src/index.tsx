import * as React from "react";
import { DispatchFn, IEpicKit } from "epickit";

type EpicKitContextValue<S> = [S, DispatchFn<S>];

export const createEpicKitContext = <S, >(epicKit: IEpicKit<S>) => {
  const context = React.createContext<EpicKitContextValue<S>>([
    epicKit.state$.value,
    epicKit.dispatch,
  ]);

  return {
    useEpicKit: createEpicKitHook(context),
    EpicKitProvider: createEpicKitProvider(context, epicKit),
  }
}

const createEpicKitProvider = <S, >(
  Context: React.Context<EpicKitContextValue<S>>,
  epicKit: IEpicKit<S>,
) => (p: {
  children: React.ReactElement,
}) => {
  const [state, setState] = React.useState(epicKit.state$.value);

  React.useEffect(() => {
    const subscription = epicKit.epic$.subscribe(
      ({state: nextState}) => setState(nextState),
    );
    return () => subscription.unsubscribe();
  }, []);

  return (
    <Context.Provider value={[state, epicKit.dispatch]}>
      {p.children}
    </Context.Provider>
  );
};

const createEpicKitHook = <S, >(
  context: React.Context<EpicKitContextValue<S>>
) => (): EpicKitContextValue<S> => {
  const value = React.useContext<EpicKitContextValue<S>>(context);
  if (value === undefined) {
    throw new Error('useEpicKit must be child of EpicKitProvider')
  }
  return value;
}
