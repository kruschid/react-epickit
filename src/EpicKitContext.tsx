import * as React from "react";

import { createEpicKit, Epic } from "../lib/epickit";
import { EpicKit, IStateDispatchProps } from "./EpicKit";

export const EpicKitContext = <S, >(initialState: S, epics?: Array<Epic<S>>) => {
  const epicKit = createEpicKit(initialState, epics);
  const Context = React.createContext<IStateDispatchProps<S>>({
    state: initialState,
    dispatch: epicKit.dispatch,
  });
  return {
    Consumer: Context.Consumer,
    Provider: (p: {children: React.ReactNode}) =>
      <EpicKit epicKit={epicKit}>
        {(epicKitProps) =>
          <Context.Provider value={epicKitProps}>
            {p.children}
          </Context.Provider>
        }
      </EpicKit>,
    };
};
