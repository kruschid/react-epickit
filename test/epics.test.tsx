import * as React from "react";
import * as renderer from "react-test-renderer";
import { interval, of, timer } from "rxjs";
import { delayWhen, filter, mapTo, switchMapTo, tap } from "rxjs/operators";

import { createAction, createActionWithPayload, Epic, IAction } from "../lib/epickit";
import { Lifecycle } from "../lib/Lifecycle";
import { EpicKit } from "../src/EpicKit";
import { EpicKitContext } from "../src/EpicKitContext";

// state definition
interface IState {
  counter: number;
}
const initialState: IState = {
  counter: 1,
};

const increment: IAction<IState> = createAction<IState>((state) => ({
  ...state, counter: state.counter + 1,
}));

const addToCounter = createActionWithPayload<IState, number>((s, p) => ({
  counter: s.counter + p,
}));

const START_COUNTING = Symbol("START_COUNTING");
const INC_COUNTER = Symbol("INC_COUNTER");

const startCounting = createAction(START_COUNTING);
const incCounter = createAction<IState>(INC_COUNTER, (state) => ({
  ...state, counter: state.counter + 1,
}));

describe("state", () => {
  it("should pass initial state to children component", async () => {
    const component = renderer.create(
      <EpicKit initialState={{counter: 1}}>
        {({state}) => <span>{state.counter}</span>}
      </EpicKit>,
    );
    expect(component).toMatchSnapshot();
    component.unmount();
  });
});

describe("reducers", () => {
  it("should pass payload to reducer", () => {
    const component = renderer.create(
      <EpicKit initialState={{counter: 1}}>
        {({state, dispatch}) =>
          <div>
            {state.counter}
            <button onClick={() => dispatch(addToCounter(5))} />
          </div>
        }
      </EpicKit>,
    );
    expect(component).toMatchSnapshot();
    component.root.find((node) => node.type === "button").props.onClick();
    expect(component).toMatchSnapshot();
    component.unmount();
  });
});

describe("epics", () => {
  it("should execute subsequent actions", () => {
    const epic$: Epic<IState> = (action$) => action$.pipe(
      filter((action) => START_COUNTING === action.type),
      switchMapTo(interval(100)),
      mapTo(incCounter),
    );

    return of(renderer.create(
      <EpicKit initialState={initialState} epics={[epic$]}>
        {({state, dispatch}) =>
          <>
            <Lifecycle onCreate={() => dispatch(startCounting)} />
            <span>{state.counter}</span>
          </>
        }
      </EpicKit>,
    ))
    .pipe(
      delayWhen((component) => {
        expect(component).toMatchSnapshot(); // 1
        return timer(120);
      }),
      delayWhen((component) => {
        expect(component).toMatchSnapshot(); // 2
        return timer(100);
      }),
      delayWhen((component) => {
        expect(component).toMatchSnapshot(); // 3
        return timer(300);
      }),
      tap((component) => {
        expect(component).toMatchSnapshot(); // 6
        component.unmount();
      }),
    )
    .toPromise();
  });

  it("should pass state and dispatch via context", () => {
    const {Provider: StateProvider, Consumer: StateConsumer} = EpicKitContext<IState>(initialState);

    const Counter = () =>
      <StateConsumer>
        {({state, dispatch}) =>
          <div>
            Counter: {state.counter}
            <button onClick={() => dispatch(increment)}></button>
          </div>
        }
      </StateConsumer>;

    const component = renderer.create(
      <StateProvider>
        <Counter />
      </StateProvider>,
    );
    expect(component).toMatchSnapshot(),
    component.root.find((node) => node.type === "button").props.onClick();
    expect(component).toMatchSnapshot(),
    component.unmount();
  });
});
