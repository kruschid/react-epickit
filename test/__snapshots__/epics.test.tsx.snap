// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`epics should execute subsequent actions 1`] = `
<span>
  1
</span>
`;

exports[`epics should execute subsequent actions 2`] = `
<span>
  2
</span>
`;

exports[`epics should execute subsequent actions 3`] = `
<span>
  3
</span>
`;

exports[`epics should execute subsequent actions 4`] = `
<span>
  6
</span>
`;

exports[`epics should pass dispatch via context 1`] = `
<div>
  Counter: 
  1
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`epics should pass dispatch via context 2`] = `
<div>
  Counter: 
  2
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`epics should pass state via context 1`] = `"1"`;

exports[`reducers should pass payload to reducer 1`] = `
<div>
  1
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`reducers should pass payload to reducer 2`] = `
<div>
  6
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`state should pass initial state to children component 1`] = `
<span>
  1
</span>
`;
